OpenKeychain/src/main/java/org/sufficientlysecure/keychain/operations/ImportExportOperation.java

```diff
@@ -590,4 +629,245 @@ ExportResult exportKeyRings(OperationLog log, long[] masterKeyIds, boolean expor
 
     }
 
-}
+    @Override
+    public OperationResult execute(ImportExportParcel input, CryptoInputParcel cryptoInput) {
+        if (input instanceof ExportKeyringParcel) {
+            ExportKeyringParcel exportInput = (ExportKeyringParcel) input;
+            switch (exportInput.mExportType) {
+                case UPLOAD_KEYSERVER: {
+                    HkpKeyserver hkpKeyserver = new HkpKeyserver(exportInput.mKeyserver);
+                    try {
+                        CanonicalizedPublicKeyRing keyring
+                                = mProviderHelper.getCanonicalizedPublicKeyRing(
+                                exportInput.mCanonicalizedPublicKeyringUri);
+                        uploadKeyRingToServer(hkpKeyserver, keyring);
+                        // TODO: replace with proper log
+                        return new ExportResult(ExportResult.RESULT_OK, new OperationLog());
+                    } catch (Exception e) {
+                        // TODO: Implement better exception handling, replace with log
+                    }
+                    break;
+                }
+                case EXPORT_FILE: {
+                    return exportToFile(exportInput.mMasterKeyIds, exportInput.mExportSecret,
+                            exportInput.mOutputFile);
+                }
+                case EXPORT_URI: {
+                    return exportToUri(exportInput.mMasterKeyIds, exportInput.mExportSecret,
+                            exportInput.mOutputUri);
+                }
+                default: {
+                    return null;
+                }
+            }
+        }
+        else if (input instanceof ImportKeyringParcel) {
+            ImportKeyringParcel importInput = (ImportKeyringParcel) input;
+            return importKeys(importInput.mKeyList, importInput.mKeyserver);
+        } else {
+            throw new RuntimeException("Invalid input parcel at ImportExportOperation");
```

Comment:
import and export should be two independent operations, to avoid control flow like this